stream-chat-android-compose/src/main/java/io/getstream/chat/android/compose/ui/attachments/StreamAttachmentFactories.kt

```diff
@@ -0,0 +1,51 @@
+package io.getstream.chat.android.compose.ui.attachments
+
+import androidx.compose.runtime.Composable
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.compose.state.messages.attachments.AttachmentState
+
+public object StreamAttachmentFactories {
+
+    /**
+     * Default attachment factories we provide, which can transform image, file and link attachments.
+     *
+     * Uses the functions below to display the UI.
+     * */
+    public val defaultFactories: List<AttachmentFactory> = listOf(
+        LinkAttachmentFactory(),
+        ImageAttachmentFactory(),
+        FileAttachmentFactory()
+    )
+}
+
+/**
+ * Holds the information required to build an attachment message.
+ * */
+public abstract class AttachmentFactory {
+    /**
+     * Returns if this specific factory can handle a specific message.
+     *
+     * @param message - The message to check.
+     * @return a boolean value, if we can consume the message and render UI.
+     * */
+    public fun canHandle(message: Message): Boolean {
+        return message.attachments.isNotEmpty() && predicate(message)
```

Comment:
I'm thinking more and more that this should be the responsibility of the factory.

Checks like `attachments.any { it.type = "whatevsies" }` will automatically be `false` for messages that don't have collections.

And we can amend checks like `attachments.all { it.type = "type" }` with an `attachments.isNotEmpty() &&` at the start. That's part of the logic of whether the factory can render the given message's attachments.

(And perhaps on the use site we can also have an extra `if` that just doesn't check in with the factories if a message has no attachments.)

Reply:
Yeah I originally wanted for our users not to have to worry about this check, but if they implement a custom factory, it's sensible to expect and `isEmpty()` check or whatever makes sense for them.

I think it would be a better approach to try and find `all` factories that work for a given message and then just render them all.